### PR TITLE
Namespace: uniform namespace pve_lxc_* and pve hostname as an API variable

### PR DESCRIPTION
**Pull request by ulvida**
I define a new uniform namespace for their role's variables, all starting with pve_lxc_, preserving backwards compatibility with previous [roles interface](https://github.com/UdelaRInterior/ansible-role-proxmox-create-lxc/blob/v2.2.0/README.md#role-variables).

Now it's possible to define `pve_lxc_hostname` in the varialbles of an installed host, to define pve hostname.

### CODE REVIEW

**Comment by vamgnu** [comment]
As this change is not backward compatible, maybe is time to change this variables:
- `pve_lxc_root_password`
- `pve_lxc_password_deploy`
to this:
- `pve_lxc_password_root`
- `pve_lxc_password_deploy`

What do you say?

**Comment by ulvida** [comment]
> As this change is not backward compatible, maybe is time to change this variables:
> 
>     * `pve_lxc_root_password`
> 
>     * `pve_lxc_password_deploy`
>       to this:
> 
>     * `pve_lxc_password_root`
> 
>     * `pve_lxc_password_deploy`
> 
> 
> What do you say?
It [was changed](https://github.com/UdelaRInterior/ansible-role-proxmox-create-lxc/blob/025583b1e1f42ec8932f28407dd82f73530d6044/defaults/main.yml#L17) to: 
* `pve_lxc_node_deploy_password`
* `pve_lxc_root_password`

Better the genitif name before ("root's password, "the password of root").

**Comment by vamgnu** [comment]
> > As this change is not backward compatible, maybe is time to change this variables:
> > ```
> > * `pve_lxc_root_password`
> > 
> > * `pve_lxc_password_deploy`
> >   to this:
> > 
> > * `pve_lxc_password_root`
> > 
> > * `pve_lxc_password_deploy`
> > ```
> > 
> > 
> > What do you say?
> > It [was changed](https://github.com/UdelaRInterior/ansible-role-proxmox-create-lxc/blob/025583b1e1f42ec8932f28407dd82f73530d6044/defaults/main.yml#L17) to:
> 
>     * `pve_lxc_node_deploy_password`
> 
>     * `pve_lxc_root_password`
> 
> 
> Better the genitif name before ("root's password, "the password of root").

To me is more practical change the suffix. But, as you wish...

**Comment by ulvida** [comment]
> As this change is not backward compatible, maybe is time to change this variables:
> 
>     * `pve_lxc_root_password`
>     * `pve_lxc_password_deploy`
>       to this:
> 
>     * `pve_lxc_password_root`
>     * `pve_lxc_password_deploy`
> 
> What do you say?

I see your point: I missed to change these variables (moreover the prefix) in all our playbooks. 

And we must point out that they didn't fail, because there are default values defined for the appropriate variables. So: is it appropriate to give default values to these passwords? I think it's  better the playbook fails till they have secure passwords defined by user than silently creating an unsecure virtual host.

Or better: we can check the strongness of the password provided and stop with a warning if they are weak.

**Comment by ulvida** [comment]
> As this change is not backward compatible, maybe is time to change this variables:
> 
>     * `pve_lxc_root_password`
>     * `pve_lxc_password_deploy`
>       to this:
> 
>     * `pve_lxc_password_root`
>     * `pve_lxc_password_deploy`
> 
> What do you say?
In fact `password_deploy` in our present playbooks is not part of this role API. This is a refactor of our role. 

Existing password variable in the role are `api_password` the password of the user wich accesses the proxmox API in the node and `root_password`, the passowrd of root user of the container. No user deploy in the container here, `node_deploy_password` is an unused variable. 

I'm wondering th use a mix namespace `pve_*` for variables shared by several proxmox modules. and  `pve_lxc_*` for specific variables to this module.

**Comment by vamgnu** [comment]
> > As this change is not backward compatible, maybe is time to change this variables:
> > ```
> > * `pve_lxc_root_password`
> > * `pve_lxc_password_deploy`
> >   to this:
> > 
> > * `pve_lxc_password_root`
> > * `pve_lxc_password_deploy`
> > ```
> > 
> > 
> > What do you say?
> > In fact `password_deploy` in our present playbooks is not part of this role API. This is a refactor of our role.
> 
> Existing password variable in the role are `api_password` the password of the user wich accesses the proxmox API in the node and `root_password`, the passowrd of root user of the container. No user deploy in the container here, `node_deploy_password` is an unused variable.
> 
> I'm wondering th use a mix namespace `pve_*` for variables shared by several proxmox modules. and `pve_lxc_*` for specific variables to this module.

You are right! I was confused things.
I think it's more appropriate use common names to common purpose. For example: credentials of API are independent if it's an lxc or kvm. Just like you say.

**Comment by ulvida** [comment]
> > > As this change is not backward compatible, maybe is time to change this variables:
> > > ```
> > > * `pve_lxc_root_password`
> > > * `pve_lxc_password_deploy`
> > >   to this:
> > > 
> > > * `pve_lxc_password_root`
> > > * `pve_lxc_password_deploy`
> > > ```
> > > 
> > > 
> > > What do you say?
> > > In fact `password_deploy` in our present playbooks is not part of this role API. This is a refactor of our role.
> > 
> > 
> > Existing password variable in the role are `api_password` the password of the user wich accesses the proxmox API in the node and `root_password`, the passowrd of root user of the container. No user deploy in the container here, `node_deploy_password` is an unused variable.
> > I'm wondering the use a mix namespace `pve_*` for variables shared by several proxmox modules. and `pve_lxc_*` for specific variables to this module.
> 
> You are right! I was confused things.
> I think it's more appropriate use common names to common purpose. For example: credentials of API are independent if it's an lxc or kvm. Just like you say.

In theory that's wonderful, but in practice, there are nuances: 
* `pve_api_host`: clearly common purpose, 
* `pve_kvm_net_interfaces` and `pve_lxc_net_interfaces` : clearly different variable structures, so different name
* but `pve_kvm_root_password` and `pve_lxc_root_password` are quite the same (the root password of the lxc or kvm) but set up in different ways or different places, specially if we clone the VM.

File: defaults/main.yml [hunk 1]
```diff
@@ -2,28 +2,51 @@
...
-memory: 512
-# swap: 512             # Optional.
-disk: 32
-storage: local-lvm
-nameserver: 192.168.8.8 192.168.8.4
+pve_hostname: "{{ inventory_hostname.split('.')[0] }}"
+# By default, we suppose that `inventory_hostname` is the FQDN or the hostname of the host to create, so we set the variable to the hostname. 
+# You can arbitrarly define this hostname
 
+# node: my_node
```
Comment (santiagomr) [inline defaults/main.yml]: I understand the intention to give a transition period before removing backwards compatibility.

However, thinking about it a bit, the result of checking if the deprecated variables are defined, is a much more complex and "dirty" code, especially for those who come to the role in this version (v3.0.0). Isn't KISS better? On the other hand, it could cause an unwanted value to be taken by the role because someone forgot to update a definition with higher precedence.

In addition, a new major version is being released precisely here, so those who adopt it will already be aware that there are disruptive changes that will involve reviewing and updating the code of their own playbooks. Or else, they will remain using the previous version.

In the absence of a precise roadmap drawn up to release v4.0.0, my opinion is that it is more convenient to make the leap right here.

Comment (ulvida) [reply defaults/main.yml]: Maybe you are right, and it's not a good idea to maintain API backwards compatibility. Facilitate (accelerate) code evolution versus complexities and superfluous patches which breake harmony of poetry... The question is valid, and I respect you prefer a cleaner code. 

I did that already for the [backupc_client role v2](https://github.com/UdelaRInterior/ansible-backuppc-client/blob/master/CHANGELOG.md), as a kind of geek exercice, and considering we had a lot of hosts using this role. So: how to merge the new version in the requirements.yml of master? With backwards compatibility, we can set the new version in master to start using it in new hosts, without going in the whole refactor, with each particular case, and letting each devop guy doing her job.

I don't think the code for this is so "complex and dirty": all is in `defaults/main.yml` and its quite clear (it even helps to update your API) and I don't see this risk: 
> it could cause an unwanted value to be taken by the role because someone forgot to update a definition with higher precedence.

Again: all is in `defaults/main.yml`, so as far as you define the new variable (which, with the new namespace, has much more chances to be unique) somewhere for the host, this value will be taken for the role, no matter what happens in defaults and/or the old API. 

Concerning your remark: 
> ... so those who adopt it will already be aware that there are disruptive changes that will involve reviewing and updating the code of their own playbooks.

Well, it's exactly what motivated this approach: I saw that many branches that involved heavier code refactor were delayed to be merged in master, breaking the: "release early, release often", and making the work harder while time goes by.

And I wouldn't do that for any role, only for these kind of roles that are deployed on many hosts, like `backuppc_client` or proxmox hosts creation. I didn't preserve the API backwards compatibility, for instance, for [backuppc server role](https://github.com/UdelaRInterior/ansible-backuppc), that was enhanced and refactored simultaneously: you usually have one, eventually a few backuppc server instances. So when you adopt the new major version of the role, you have to work it out. And there, you should start using the new major version and API of the backuppc client role. But then, you need a roadmap to deploy the new API in all your clients. The backwards compatibility of the API pretends to help in this deployment.

But well: in both cases, backuppc and proxmox_create, finally I updated the API of all the clients myself before merging... so it's not really useful (except to have two major version updates, which may make our roles to seem "mature" :-) .

Comment (ulvida) [reply defaults/main.yml]: Concerning the roadmap, simple: the API is updated in our playbooks' set is, so as soon as we test this version, we can publish next major version, just dropping this backward's compatibility.

Comment (vamgnu) [reply defaults/main.yml]: For comment something: a default value is just a default value. For someone who recently arrived to the code it's confused.
I think that ansible is not a programming language, is a way to describe our infrastructure as code.
These changes is not KISS...

Comment (santiagomr) [reply defaults/main.yml]: >  how to merge the new version in the requirements.yml of master?

Merging this new version in our particular IaC project will be the same in either of the two cases, the refactor will be done anyway, we just have to decide whether to do it all at once or extend it over time by making new hosts coexist in the recent structure with old hosts that may or may not have been upgraded by a devop guy.

For branches that have delayed behind *master*, its requirements.yml will continue pointing to the v2.x.y of the role so consistency and integrity will be maintained.

> I don't think the code for this is so "complex and dirty"

I consider that it is for those who wish to use the role from this new version and have no idea or interest in the structure and names of previous variables.

> and I don't see this risk

You are right! After reviewing it again, the risk of taking unwanted values is practically nil because it is precisely in `defaults/main.yml` as you say.

In conclusion, both ways are equally valid. It's a matter of choosing the one that best suits.

I give my vote to KISS

Comment (ulvida) [reply defaults/main.yml]: Let's do it simple, without re-writing history: in our IaC, we no longer need this backward's compatibility, if ever we did. So let's prepare and publish the next major version, dropping this questioned code. Do you both agree?

I started [for proxmox_create_lxc](https://github.com/UdelaRInterior/ansible-role-proxmox-create-lxc/pull/12/commits/a4d01fbf74322641f9a7f489668774d3457ca966), tested with our host `pavada`.

Comment (ulvida) [reply defaults/main.yml]: And I did it also [for KVM role](https://github.com/UdelaRInterior/ansible-role-proxmox-create-kvm/pull/9). 

Both basically tested with an LXC and a KVM creation.

Comment (vamgnu) [reply defaults/main.yml]: > tested with our host `pavada`.

Don't you think that is a "`pavada`", don't you? :-)

**Comment by andrespias** [comment]
I'm late ... but I read everything. good work! and interesting discussion!

File: defaults/main.yml (hunk 1 above)
Comment (ulvida) [reply defaults/main.yml]: In French we would say: "Les doigts dans le nez!" (_Finger in the nose!_), that could be translated by: "Una pavda!" ...and in the US you could do that for Thanksgiving.

**Review by santiagomr — APPROVED** [review]
Resolved in discussion

**Comment by santiagomr** [comment]
From the changes added and discussed in this unconcluded PR, another branch was created to make new updates and refactorings to the role. Therefore this PR is truncated and replaced by [this other](https://github.com/UdelaRInterior/ansible-role-proxmox-create-lxc/pull/14)

**Comment by ulvida** [comment]
> From the changes added and discussed in this unconcluded PR, another branch was created to make new updates and refactorings to the role. Therefore this PR is truncated and replaced by [this other](https://github.com/UdelaRInterior/ansible-role-proxmox-create-lxc/pull/14)

But why not merging the branch of this PR into master and start the new PR from there? Is there something still to improve in this code? I agree that we must move to ansible 2.10 and over, but it's also important to bring our improved roles to our production IaS code.

**Comment by santiagomr** [comment]
Because when finalizing the review and merge of the [new PR](https://github.com/UdelaRInterior/ansible-role-proxmox-create-lxc/pull/14), thise PR changes and their tags will also be included, so the history will be respected and maintained.

However, the code proposed in this PR should not be being used since it never gave up staging. And much of the imminent release of PR #14 contains refactorings to variables that are not backward compatible with this intermediate refactoring.

In conclusion, merging this PR is like releasing a code that already has a replacement.
